interceptor.go

```diff
@@ -142,78 +63,66 @@ func (i *interceptor) WrapUnary(next connect.UnaryFunc) connect.UnaryFunc {
 		}
+		carrier := propagation.HeaderCarrier(request.Header())
+		spanCtx := trace.SpanContextFromContext(ctx)
+		ctx = trace.ContextWithRemoteSpanContext(ctx, spanCtx)
+		ctx = i.config.propagator.Extract(ctx, carrier)
+		ctx, span := i.config.tracer.Start(
+			ctx,
+			name,
+			trace.WithSpanKind(spanKind),
+		)
 		i.config.propagator.Inject(ctx, carrier)
 		defer span.End()
-		reqSpanType, resSpanType := semconv.MessageTypeKey.String("RECEIVED"), semconv.MessageTypeKey.String("SENT")
-		if request.Spec().IsClient {
-			reqSpanType, resSpanType = resSpanType, reqSpanType
-		}
-		reqSpanAttrs := []attribute.KeyValue{reqSpanType, semconv.MessageIDKey.Int(1)}
-		if msg, ok := request.Any().(proto.Message); ok {
-			size := proto.Size(msg)
-			reqSpanAttrs = append(reqSpanAttrs, semconv.MessageUncompressedSizeKey.Int(size))
-		}
-		span.AddEvent("message", trace.WithAttributes(reqSpanAttrs...))
+		requestSize := msgSize(request)
+		span.AddEvent(messageKey,
+			trace.WithAttributes(
+				requestSpan,
+				semconv.MessageIDKey.Int(1),
+				semconv.MessageUncompressedSizeKey.Int(requestSize),
+			),
+		)
 		response, err := next(ctx, request)
-		codeAttr := statusCodeAttribute(protocol, err)
-		attrs = append(attrs, codeAttr)
-		if msg, ok := request.Any().(proto.Message); ok {
-			size := proto.Size(msg)
-			instr.requestSize.Record(ctx, int64(size), attrs...)
-		}
-		resSpanAttrs := []attribute.KeyValue{resSpanType, semconv.MessageIDKey.Int(1)}
-		if err == nil {
-			if msg, ok := response.Any().(proto.Message); ok {
-				size := proto.Size(msg)
-				instr.responseSize.Record(ctx, int64(size), attrs...)
-				resSpanAttrs = append(resSpanAttrs, semconv.MessageUncompressedSizeKey.Int(size))
-			}
-		}
-		span.AddEvent("message", trace.WithAttributes(resSpanAttrs...))
-		instr.duration.Record(ctx, i.config.now().Sub(requestStartTime).Milliseconds(), attrs...)
-		instr.requestsPerRPC.Record(ctx, 1, attrs...)
-		instr.responsesPerRPC.Record(ctx, 1, attrs...)
-		if err != nil {
-			if connectErr := new(connect.Error); errors.As(err, &connectErr) {
-				span.SetStatus(codes.Error, connectErr.Message())
-			} else {
-				span.SetStatus(codes.Error, err.Error())
-			}
-		}
-		span.SetAttributes(codeAttr)
+		attributes = append(attributes, statusCodeAttribute(protocol, err))
+		responseSize := msgSize(response)
+		span.AddEvent(messageKey,
+			trace.WithAttributes(
+				responseSpan,
+				semconv.MessageIDKey.Int(1),
+				semconv.MessageUncompressedSizeKey.Int(responseSize),
+			),
+		)
+		span.SetStatus(spanStatus(err))
+		span.SetAttributes(attributes...)
+		instrumentation.duration.Record(ctx, i.config.now().Sub(requestStartTime).Milliseconds(), attributes...)
+		instrumentation.requestSize.Record(ctx, int64(requestSize), attributes...)
+		instrumentation.requestsPerRPC.Record(ctx, 1, attributes...)
+		instrumentation.responseSize.Record(ctx, int64(responseSize), attributes...)
+		instrumentation.responsesPerRPC.Record(ctx, 1, attributes...)
```

Comment:
trace/spans will always be evaluated before metrics. 
Improves readability a lot :)

Reply:
Makes sense. Especially in big diffs, it's nice for the reviewer to have trivial and significant changes in separate commits - reordering doesn't usually require the same attentive review as, for example, the mutex scoping changes here.